proposals/NNNN-extensible-enums.md

```diff
@@ -0,0 +1,208 @@
+  tricky. For example, `SwiftNIO` uses an enumeration to represent HTTP status
+  codes. If new status codes are added, SwiftNIO needs to either mint new
+  enumerations and do a deprecate-and-replace, or it needs to force these new
+  status codes through the .custom enum case.
+
+This proposal plans to address these limitations on enumerations in
+non-resilient Swift.
+
+## Proposed solution
+
+We propose to introduce a new language feature `ExtensibleEnums` that aligns the
+behaviour of enumerations in both language dialects. This will make **public**
+enumerations in packages a safe default and leave maintainers the choice of
+extending them later on.
+
+In modules with the language feature enabled, developers can use the existing
+`@frozen` attribute to mark an enumeration as non-extensible, allowing consumers
+of the module to exhaustively switch over the cases. This makes committing to the
+API of an enum an active choice for developers.
+
+Modules consuming other modules with the language feature enabled will be forced
+to add an `@unknown default:` case to any switch state for enumerations that are
+not marked with `@frozen`. Importantly, this only applies to enums that are
+imported from other modules that are not in the same package. For enums inside
+the same modules of the declaring package switches is still required to be
+exhaustive and doesn't require an `@unknown default:` case.
+
+Since enabling a language feature applies to the whole module at once we also
+propose adding a new attribute `@extensible` analogous to `@frozen`. This
+attribute allows developers to make a case-by-case decision on each enumeration
+if it should be extensible or not by applying one of the two attributes. The
+language feature `ExtensibleEnums` can be thought of as implicitly adding
+`@extensible` to all enums that are not explicitly marked as `@frozen`.
+
+In resilient modules, the `@extensible` attribute doesn't affect API nor ABI
+since the behaviour of enumerations in modules compiled with library evolution
+mode are already extensible by default. We believe that extensible enums are the
+right default choice in both resilient and non-resilient modules and the new
+proposed `@extensible` attribute primiarly exists to give developers a migration
+path.
+
+## Source compatibility
+
+Enabling the language feature `ExtensibleEnums` in a module that contains public
+enumerations is a source breaking change.
+Changing the annotation from `@frozen` to `@extensible` is a source breaking
+change. 
+Changing the annotation from `@extensible` to `@frozen` is a source compatible
+change and will only result in a warning code that used `@unknown default:`
+clause. This allows developers to commit to the API of an enum in a non-source
+breaking way.
+Adding an `@extensible` annotation is a source breaking change in modules that
+have **not** enabled the `ExtensibleEnums` language features or are compiled
+with resiliency.
```

Comment:
It might be worth clarifying that it's only a source breaking change if `@extensible` is added to an _existing_ public enum?